platform/kube/register.go

```diff
@@ -0,0 +1,196 @@
+		if !found {
+			r.Name = str
+		}
+	}
+	return r, nil
+}
+
+// samePorts returns true if the numerical part of the ports is the same.
+// The arrays aren't necessarily sorted so we (re)use a map.
+func samePorts(ep []v1.EndpointPort, portsMap map[int32]bool) bool {
+	if len(ep) != len(portsMap) {
+		return false
+	}
+	for _, e := range ep {
+		if !portsMap[e.Port] {
+			return false
+		}
+	}
+	return true
+}
+
+// splitEqual splits key=value string into key,value. if no = is found
+// the whole string is the key and value is empty.
+func splitEqual(str string) (string, string) {
+	idx := strings.Index(str, "=")
+	var k string
+	var v string
+	if idx >= 0 {
+		k = str[:idx]
+		v = str[idx+1:]
+	} else {
+		k = str
+	}
+	return k, v
+}
+
+// addLabelsAndAnnotations adds labels and annotations to an object.
+func addLabelsAndAnnotations(obj *meta_v1.ObjectMeta, labels []string, annotations []string) {
+	if obj.Labels == nil {
+		obj.Labels = make(map[string]string, len(labels))
+	}
+	for _, l := range labels {
+		k, v := splitEqual(l)
+		obj.Labels[k] = v
+	}
+	if obj.Annotations == nil {
+		obj.Annotations = make(map[string]string, len(annotations))
+	}
+	for _, a := range annotations {
+		k, v := splitEqual(a)
+		obj.Annotations[k] = v
+	}
+}
+
+// RegisterEndpoint registers the endpoint (and the service if it doesn't
+// already exists). It creates or updates as needed. When creating it adds the
+// optional labels.
+func RegisterEndpoint(client kubernetes.Interface, namespace string, svcName string,
+	ip string, portsList []NamedPort, labels []string, annotations []string) error {
+	getOpt := meta_v1.GetOptions{IncludeUninitialized: true}
+	_, err := client.Core().Services(namespace).Get(svcName, getOpt)
+	if err != nil {
+		glog.Warningf("Got '%v' looking up svc '%s' in namespace '%s', attempting to create it", err, svcName, namespace)
+		svc := v1.Service{}
+		svc.Name = svcName
+		for _, p := range portsList {
+			svc.Spec.Ports = append(svc.Spec.Ports, v1.ServicePort{Name: p.Name, Port: p.Port})
+		}
+		addLabelsAndAnnotations(&svc.ObjectMeta, labels, annotations)
+		_, err = client.CoreV1().Services(namespace).Create(&svc)
+		if err != nil {
+			glog.Error("Unable to create service: ", err)
+			return err
+		}
+	}
+	eps, err := client.CoreV1().Endpoints(namespace).Get(svcName, getOpt)
+	if err != nil {
+		glog.Warningf("Got '%v' looking up endpoints for '%s' in namespace '%s', attempting to create them",
+			err, svcName, namespace)
+		endP := v1.Endpoints{}
+		endP.Name = svcName // same but does it need to be
+		addLabelsAndAnnotations(&endP.ObjectMeta, labels, annotations)
+		eps, err = client.CoreV1().Endpoints(namespace).Create(&endP)
+		if err != nil {
+			glog.Error("Unable to create endpoint: ", err)
+			return err
+		}
+	}
+	// To check equality:
+	portsMap := make(map[int32]bool, len(portsList))
+	for _, e := range portsList {
+		portsMap[e.Port] = true
+	}
+
+	glog.V(2).Infof("Before: found endpoints %+v", eps)
```

Comment:
can you atleast make these messages meaningful? To an end user debugging things before they come to mailing list, a before, after, found etc is not going to be helpful.

Reply:
it's V(2) which means high level of debugging and meant for debugging. if a user has an issue we may ask them to run the command with V(2) to get more context. I'm not sure what you mean by "more meaningful"

Reply:
What does before mean here? Some context as to what is happening during the time of printing this message would be helpful. Otherwise, you would be the only person capable of trouble shooting this. We don't go look at code everytime a user complains of an error do we?

Reply:
we can discuss debuggability separately - more logging (at the right level) is often what makes the difference between knowing what happened vs having to make a patch release to improve logging.